c++/src/kj/compat/http.c++

```diff
@@ -6402,12 +6822,35 @@ private:
 
     kj::Maybe<CompressionParameters> acceptedParameters;
     kj::String agreedParameters;
-    if (server.settings.enableWebSocketCompression) {
+    auto compressionMode = server.settings.webSocketCompressionMode;
+    if (compressionMode == HttpServerSettings::AUTOMATIC_COMPRESSION) {
+      // If AUTOMATIC_COMPRESSION is enabled, we ignore the `headers` passed by the application and
+      // strictly refer to the `requestHeaders` from the client.
       KJ_IF_MAYBE(value, requestHeaders.get(HttpHeaderId::SEC_WEBSOCKET_EXTENSIONS)) {
+        // Perform compression parameter negotiation.
         KJ_IF_MAYBE(config, _::tryParseExtensionOffers(*value)) {
           acceptedParameters = kj::mv(*config);
         }
       }
+    } else if (compressionMode == HttpServerSettings::MANUAL_COMPRESSION) {
+      // If MANUAL_COMPRESSION is enabled, we use the `headers` passed in by the application, and
+      // try to find a configuration that respects both the server's preferred configuration,
+      // as well as the client's requested configuration.
+      //
+      // TODO(now): Manual Mode and terminating in worker
```

Comment:
What if the caller of `acceptWebSocket()` simply always sets `Sec-WebSocket-Extension: permessage-deflate`? If the client didn't ask for `permessage-deflate`, then KJ will filter it out, right? So would this be an effective way to accept whatever the client offered?

If that doens't work, then I think we should expose a utility function like `kj::choosePreferredWebSocketExtensions(kj::StringPtr clientOffer)`. When a WebSocket is terminating in a worker, the Workers Runtime could call this function to construct the Sec-WebSocket-Extensions header.

Reply:
IMO, this is fundamentally an issue on the workers (application) side of things -- we're in manual mode but we aren't passing headers to `acceptWebSocket()`. I think we should determine if we're proxying, then populate `headers` (with the client's requested `Sec-WebSocket-Extensions` value) if we're terminating in a worker. That would resolve the issue.

### Terminating websocket in a worker
This is the current state of things:
1. Client offers `permessage-deflate`
2. worker gets the request, we start building the response and see we want to accept a websocket
3. `headers` passed to `acceptWebSocket()` is empty
5. Because we're in manual mode, we look at `headers` and see it doesn't want to use compression so we reply to the client without `Sec-WebSocket-Extensions` (even if the client requested it).

We don't want to set `Sec-WebSocket-Extensions: permessage-deflate` because that would prevent us from ever allowing the `server_x` parameters (and we would always ignore the `client_x` params). I think the most appropriate thing to do is just use the exact header that the client provided.
### Proxying
Proxying should work correctly, since we set `headers` to be the response headers from the origin server.